Reconcile RetrySettings & Clock

vam-google opened the pull request:
Reconcile RetrySettings in GAX with RetryParams in google-cloud-java
https://github.com/GoogleCloudPlatform/google-cloud-java/issues/1574
Reconcile NanoClock in GAX with Clock in google-cloud-java
https://github.com/GoogleCloudPlatform/google-cloud-java/issues/1575

garrettjonesgoogle commented on src/test/java/com/google/api/gax/grpc/RecordingScheduler.java:
```diff
@@ -29,41 +29,63 @@
-    clock.setCurrentNanoTime(clock.nanoTime() + TimeUnit.NANOSECONDS.convert(delay, unit));
-    return executor.schedule(runnable, 0, TimeUnit.NANOSECONDS);
-  }
+  static RecordingScheduler get(final FakeNanoClock clock) {
```
maybe call this `create` instead of `get` - it matches other classes in GAX better.

vam-google replied:
Changed

garrettjonesgoogle commented on src/test/java/com/google/api/gax/grpc/CancellationTest.java:
```diff
@@ -115,24 +120,36 @@ public void cancellationBeforeGetOnRetryingCallable() throws Exception {
-    @Override
-    public List<Runnable> shutdownNow() {
-      return executor.shutdownNow();
+  private abstract static class LatchCountDownScheduler implements ScheduledExecutorService {
```
Could you explain the reason for this change?

vam-google replied:
This allows using SchedulerExecutorService directly in UnaryCallable and oteher callables, and avoids using the UnaryCallable.Scheduler. I believe that UnaryCallable.Scheduler was introduced just because of complications with mocking the full ScheduledExecutorService. Also, I had to avoid using UnaryCallable.Scheduler, because the retry call should be Grpc-independent.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/core/RetrySettings.java:
```diff
@@ -92,6 +95,13 @@
   public abstract Duration getMaxRetryDelay();
 
   /**
+   * MaxAttempts defines the maximum number of attempts to perform. If number of attempts reaches
```
"If number of attempts" -> "If the number of attempts"

garrettjonesgoogle commented on src/main/java/com/google/api/gax/core/RetrySettings.java:
```diff
@@ -92,6 +95,13 @@
 
   /**
+   * MaxAttempts defines the maximum number of attempts to perform. If number of attempts reaches
+   * this limit the logic will give up retrying even if the total retry time is still lower than
```
"this limit the logic" -> "this limit, the logic"

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/RetryingCallable.java:
```diff
@@ -236,19 +100,61 @@ private static CallContext getCallContextWithDeadlineAfter(
-  private static boolean canRetry(Throwable throwable) {
-    if (!(throwable instanceof ApiException)) {
-      return false;
+  private static class GrpcRetryCallable<RequestT, ResponseT> implements Callable<ResponseT> {
```
maybe this should be `GrpcRetryingCallable`

shinfan commented on src/main/java/com/google/api/gax/retrying/AbstractRetryHandler.java:
```diff
@@ -0,0 +1,137 @@
+ *
+ * @param <ResponseT> response type
+ */
+public abstract class AbstractRetryHandler<ResponseT> implements RetryHandler<ResponseT> {
```
I don't see the reason of having both AbstractRetryHandler and RetryHandler. Can we just make RetryHandler an abstract class?

vam-google replied:
I agree with you, and I don't like having useless classes/interfaces either. Having a separate interface feels cleaner, and also it does not put the single-inheritance restriction on any subclasses. I'm still not sure which one is better, though.

shinfan replied:
Do we have any multiple-inheritance usage now for RetryHandler? I agree that if we need multiple-inheritance we can add this interface later. Otherwise, there is no need to add this extra interface now.

vam-google commented on src/main/java/com/google/api/gax/core/DefaultNanoClock.java:
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
```
This was not created but moved to another package, so I kept the date as is. Should I update it for moved files too?

garrettjonesgoogle replied:
Ahh in that case, keep the year.

vam-google commented on src/main/java/com/google/api/gax/core/SystemClock.java:
```diff
@@ -0,0 +1,58 @@
+import java.io.Serializable;
+import java.util.concurrent.TimeUnit;
+
+public final class SystemClock implements NanoClock, Serializable {
```
This clock is used in gcj repo, where it is a field in various serializable classes (so supposed to be serializable too). Also there are several serialization tests, and the clock is part of those tests too.

vam-google commented on src/main/java/com/google/api/gax/retrying/AbstractRetryHandler.java:
```diff
@@ -0,0 +1,137 @@
+ *
+ * @param <ResponseT> response type
+ */
+public abstract class AbstractRetryHandler<ResponseT> implements RetryHandler<ResponseT> {
```
What exactly do you mean? It is possible to not use an abstract class, but the current structure (with the class as is) ended up the cleanest solution I had.

vam-google commented on src/main/java/com/google/api/gax/retrying/AbstractRetryHandler.java:
```diff
@@ -0,0 +1,137 @@
+    return totalTimeSpentNanos <= totalTimeoutNanos
+        && (globalSettings.getMaxAttempts() <= 0
+            || nextAttemptSettings.getAttemptCount() < globalSettings.getMaxAttempts());
+  }
```
I tried to keep the RetryAttemptSettings class as a POJO without any logic in it (similar to the RetrySettings). It makes sense, as the attemptSettings represent the settings themselves, which can be interpreted differently, depending on context/client/type of client.

garrettjonesgoogle replied:
My suggestion implied a name change too (although I wasn't clear about that). I was trying to get at a more central point: I think there are two separable concepts here - the logic for determining when/if to retry, and the logic for executing that retry with the right timing. Also, if we separate the two concepts, we won't need an abstract class.

vam-google replied:
I think there are 3 logics here:
1) Logic determining if we should retry. It is based on:
    1.1) Previous values
    1.2) Current time (not directly related to the settings, should be passed as parameter)
    1.3) Previous attempt exception (not related to the settings either, should be passed as parameter). Also exception checking is not as simple as just checking for exception type or returned code, just look at the ExceptionHandler in gcj, which is 300 lines of pretty sophisticated stuff.
2) Determining what the next attempt settings should be (because of that case, when we retry immediately if grpc deadline is exceeded, that logic is conditional, and is different for rgpc and non-grpc)
3) Executing the attempt.

And here the last one is the simplest part. If I take everything except the execution from RetryHandler and move it to RetryAttemptSettings, then RetryAttemptSettings will practically become what RetryHandler is now, but with a different name. 

Having a separate class for execution only felt like an overkill (but still cannot easily use ScheduledExecutorService directly, because we need a direct (same thread) execution, and there is no version of MoreExecutors.directExecutor() for Scheduled executor).

Also, please note, that some of these design decision were dictated by the gcj integration part. Please check how this handler is used in the related gcj pull request.

Please clarify how strong you are about making this change. I would still prefer not doing it, but if you feel really strong about it, I'm fine with making the change (in that case, please suggest names for the RetryAttemtSettings, as CallAttemptSettings does not feel like a class with lots of logic in it). Note, it will require changes in gcj integration too (basically having something like gax and gcj AttemptSettings, which I believe defeats the purpose of the whole change)

vam-google commented on src/main/java/com/google/api/gax/retrying/AbstractRetryHandler.java:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  /**
+   * Crates next attempt settings. It increments the current attempt count and uses randomized
```
Corrected

vam-google commented on src/main/java/com/google/api/gax/retrying/AbstractRetryHandler.java:
```diff
@@ -0,0 +1,137 @@
+        Duration.millis(ThreadLocalRandom.current().nextLong(newRetryDelay)),
+        prevSettings.getAttemptCount() + 1,
+        prevSettings.getFirstAttemptStartTime());
+  }
```
Please check the previous comment. I really dislike putting logic into those type of classes. That class contains only configuraiton parameters of a concrete retry and nothing more, the name settings usually implies that, and is consistent with the rest of code (for example with RetrySettings).

vam-google commented on src/main/java/com/google/api/gax/retrying/RetryAttemptSettings.java:
```diff
@@ -0,0 +1,106 @@
+/**
+ * Execution attempt settings. Defines attempt-specific properties of a retry process.
+ */
+public class RetryAttemptSettings {
```
I'm fine with renaming it to CallAttemptSettings, or even just AttemptSettings. I called it RetryAttemptSettings, because without that prefix it is not clear what "attempt" is even about. The "Call" or any other prefix - i did not want to introduce any "new" terms into the process to keep things simple.

garrettjonesgoogle replied:
If it's used for the first call, then `CallAttemptSettings` is better. We use `Call` all over the place.

vam-google commented on src/test/java/com/google/api/gax/grpc/RecordingScheduler.java:
```diff
@@ -29,41 +29,63 @@
-  private final ScheduledExecutorService executor;
-  private final List<Duration> sleepDurations = new ArrayList<>();
-  private final FakeNanoClock clock;
+abstract class RecordingScheduler implements ScheduledExecutorService {
```
Because otherwise I would have to implement 17 other methods in ScheduledExecutorService interface, which are irrelevant for our test. I refactored this class, it used to be implementing UnaryCallable.Scheduler instead, which was a wrapper interface. I believe that interface existed only because of the problems with mocking the full scheduler for the testing purposes. 

Now, when I think about it, it does not even have to be a class (can be just an interface, since the implementation is still completely generated  Mockito during runtime). I will change it to that, if it is ok for you.

garrettjonesgoogle replied:
Yeah I think that's right. It does seem better to go with the mocking approach.

vam-google commented on src/main/java/com/google/api/gax/retrying/RetryHandler.java:
```diff
@@ -0,0 +1,101 @@
+      Callable<ResponseT> callable, RetrySettings globalSettings);
+
+  /**
+   * Creates next attempt {@link RetryAttemptSettings}, which defines properties of the next
```
Changed

vam-google commented on src/main/java/com/google/api/gax/retrying/RetryHandler.java:
```diff
@@ -0,0 +1,101 @@
+   * @param attemptSettings current attempt settings
+   * @return the {@link Future}, representing the scheduled execution
+   */
+  Future<ResponseT> executeAttempt(
```
Because it does not return ApiFuture. In typical case it returns whatever Future is returned by the underlining ShceduledExecutorService. I can wrap it in ApiFuture, but trying to avoid wrappers when they are not necessary (there are already wrappers around everything, including wrappers around wrappers).

garrettjonesgoogle replied:
The `ApiFuture` interface provides the callback hooks, which is important for creating a performant application. I'd like to make it a convention to return it everywhere to make the callbacks possible.

vam-google replied:
That future is for internal use. I would rather discourage users from assigning listeners to that particular future, since it can be misleading and it is already very hard to provide correct execution guarantees for that future and its listeners. 

For example, for the grpc case, the there can be 2 different active attempt futures behind RetryFuture (not simultaneously): 1) the scheduled within ScheduledExecutorService; 2) or the actual FutureCallable result (it replaces RetryFuture's internal attempt future, once it is time to actually do the new attempt, and the attempt itself is an asynchronous call). 

In other words, assiging listeners to the future returned by the executeAttempt method, will most probably not be called when user expects them to be called. 

I'm agree with having ApiFuture whenever possible. But, here, when it is internal use only, and when we already have the complex logic with futures (proxying, substitution, cancelation etc) I would like to keep the interface as generic as possible (so it is harder to mess up something).

vam-google commented on src/main/java/com/google/api/gax/retrying/DirectRetryHandler.java:
```diff
@@ -0,0 +1,81 @@
+import org.joda.time.Duration;
+
+/**
+ * The retry handler, which executes attempts in the current thread, potentially causing the current
```
Corrected

vam-google commented on src/main/java/com/google/api/gax/retrying/DirectRetryHandler.java:
```diff
@@ -0,0 +1,81 @@
+      if (Duration.ZERO.compareTo(attemptSettings.getRandomizedRetryDelay()) < 0) {
+        Thread.sleep(attemptSettings.getRandomizedRetryDelay().getMillis());
+      }
+      return Futures.immediateFuture(callable.call());
```
Not sure if I understand correctly what you mean, but triggering sleep on Future.get(), or any other scheduling based on Future.get() call would make an assumption that Future.get() is/will be called at all. This would be a wrong assumption, because not all calls even return a value and/or are awaited for completion.

garrettjonesgoogle replied:
Right - it was probably an impossible idea, but I thought I'd throw it out there.

vam-google commented on src/main/java/com/google/api/gax/retrying/RetryFutureImpl.java:
```diff
@@ -0,0 +1,175 @@
+ *
+ * <ul>
+ * <li>Schedules next attempt in case of failure using callback chaining technique.</li>
+ * <li>Terminates retrying process if more retries are not accepted.</li>
```
"Retrying" seems much more descriptive here, doesn't it?

vam-google commented on src/main/java/com/google/api/gax/core/RetrySettings.java:
```diff
@@ -92,6 +95,13 @@
   /**
+   * MaxAttempts defines the maximum number of attempts to perform. If number of attempts reaches
+   * this limit the logic will give up retrying even if the total retry time is still lower than
+   * TotalTimeout.
```
Added

vam-google commented on src/main/java/com/google/api/gax/grpc/RetryingCallable.java:
```diff
@@ -29,198 +29,62 @@
-    return resultFuture;
+    GrpcRetryCallable<RequestT, ResponseT> retryCallable =
+        new GrpcRetryCallable<>(callable, request, context);
+    GrpcRetryHandler<ResponseT> retryHelper = new GrpcRetryHandler<>(clock, scheduler);
```
Renamed

vam-google commented on src/main/java/com/google/api/gax/grpc/RetryingCallable.java:
```diff
@@ -29,198 +29,62 @@
+    GrpcRetryCallable<RequestT, ResponseT> retryCallable =
+        new GrpcRetryCallable<>(callable, request, context);
+    GrpcRetryHandler<ResponseT> retryHelper = new GrpcRetryHandler<>(clock, scheduler);
+    RetryFuture<ResponseT> rv = retryHelper.createFirstAttempt(retryCallable, retryParams);
```
Ok. rv means "returned value", I use it often, but I agree, that it is rather a bad habit than a good practice.

vam-google commented on src/main/java/com/google/api/gax/grpc/RetryingCallable.java:
```diff
@@ -236,19 +100,61 @@ private static CallContext getCallContextWithDeadlineAfter(
+    @Override
+    public RetryAttemptSettings createNextAttemptSettings(
+        Throwable e, RetryAttemptSettings prevSettings) {
+      if (((ApiException) e).getStatusCode() == Code.DEADLINE_EXCEEDED) {
```
Yes, because otherwise it would be rejected by the accept() method of same class earlier (not making the check for the second time keeps things in a fail-fast mode).

vam-google commented on src/main/java/com/google/api/gax/retrying/RetryFutureImpl.java:
```diff
@@ -0,0 +1,175 @@
+ * interface, and is doing the following:
+ *
+ * <ul>
+ * <li>Schedules next attempt in case of failure using callback chaining technique.</li>
```
Changed.

vam-google commented on src/test/java/com/google/api/gax/grpc/FakeNanoClock.java:
```diff
@@ -29,8 +29,9 @@
-class FakeNanoClock implements NanoClock {
+import com.google.api.gax.core.NanoClock;
 
+class FakeNanoClock implements NanoClock {
```
Agree. Moved.

vam-google commented:
Addressed review comments, PTAL.

vam-google commented on version.txt:
```diff
@@ -1 +1 @@
-0.3.1-SNAPSHOT
+0.3.2-SNAPSHOT
```
took out

garrettjonesgoogle commented on src/main/java/com/google/api/gax/retrying/RetryFutureImpl.java:
```diff
@@ -74,22 +74,28 @@
-        callbackFutureCallback.attemptFuture.cancel(mayInterruptIfRunning);
+        boolean rv = callbackFutureCallback.attemptFuture.cancel(mayInterruptIfRunning);
+        super.cancel(mayInterruptIfRunning);
+        return rv;
```
another case of `rv`

vam-google replied:
Changed, though I will miss rvs. The idea behind using generic "returned value" as a name for returned value is that it is usually difficult to come up with a reasonable name for that variable anyways, and what exactly the value is about is always explained by the name of the method itself.

shinfan commented on src/main/java/com/google/api/gax/core/RetrySettings.java:
```diff
@@ -151,6 +161,13 @@ public Builder toBuilder() {
+     * this limit the logic will give up retrying even if the total retry time is still lower than
+     * TotalTimeout.
+     */
+    public abstract Builder setMaxAttempts(int maxAttempts);
```
There is a minimum attempts field in the gcloud RetryParam class. Are we deprecating that functionality?

vam-google replied:
Yes. It was part of the original design.

vam-google commented:
PTAL

Addressed the review comments, plus some other smaller changes:
1. Split RetryHandler on 3 different components: 1) scheduling logic, 2) time-specific and 3) exception-specific retry logics.
2. Removed abstract classes.
3. Renamed NanoClock to ApiClock, as it has System.currentTimeMillis() implementation, and NanoClock name is misleading now.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/core/NanoClock.java:
```diff
@@ -29,23 +29,33 @@
+import java.util.concurrent.TimeUnit;
+
+/** Default implementation of the ApiClock interface, using call to System.nanoTime(). */
+public final class NanoClock implements ApiClock, Serializable {
```
Maybe call this `SystemNanoClock`

vam-google replied:
The other clock is called SystemClock to indicate that it based on system time (actual time, which is meaningful on any other system). In case of nano clock it can be used only to measure time difference between two events on same machine and only. I.e most important difference between two clocks is NOT the precision, but the actual meaning of the returned value.

garrettjonesgoogle replied:
But... this one calls `System.nanoTime()`... That same confusion is present on the `System` class itself. I prefer to channel the underlying concepts in this case instead of trying to "improve" them.

vam-google replied:
For SystemClock "system" is not about class name used in implementation, it is about actual system (machine) it is running on. NanoClock - just says that it has nano precision, and does not guarantee that the time is bounded to any real point in time.

Or please suggest 2 other names to those classes, but such that it is clear that SystemClock provides real world time, while NanoClock just a way to measure time difference between two events on same machine.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/core/SystemClock.java:
```diff
@@ -34,12 +34,16 @@
+ * Implementation of the {@link ApiClock} interface, which uses {@link System#currentTimeMillis()}
+ * as time source.
+ */
+public final class SystemClock implements ApiClock, Serializable {
```
Maybe call this `SystemMillisClock`

vam-google replied:
Please check the previous comment.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/retrying/RetryingFutureImpl.java:
```diff
@@ -0,0 +1,183 @@
+      }
+      if (attemptFuture != null) {
+        attemptFutureCallback = new AttemptFutureCallback(attemptFuture);
+        Futures.addCallback((ListenableFuture) attemptFuture, attemptFutureCallback);
```
You require a `ListenableFuture` to be passed into `setAttemptFuture` (otherwise there will be a `ClassCastException`), but this is not documented, and additionally, it requires usage of Guava to call this method. `ApiFuture` was designed to fix this problem: no casting is necessary, and it provides the functionality that people need to interact with this library. I am not personally concerned about the extra layers of wrapping.

garrettjonesgoogle replied:
In general: I don't want to make direct reference to future stuff in Guava except in places it's forced on us by dependencies (e.g. grpc).

vam-google replied:
Yes, it requires that future to be Guava's. It is also a package-private class and it is mostly an implementation detail. 

During its life cycle, RetryingFuture wraps 2 different types of future for each type (one after another): first it wraps the ScheduledExecutor's FutureTask, when the task is waiting for execution in some time in future, after that the RetryingFuture replaces it with ApiFuture returned by the gax.grpc's FutureCallable. One of the reasons, why I prefer keeping generic Future interface is to indicate the fact that they can be really different futures.

Main reason why I don't want more layers is my experience with guava's futures so far. They work well for 99% cases, but when it comes to corner cases they are not that good. Specifically guava usually wraps classes from `java.util.concurrent` but does it with limitations, which are essential for cases like retries. Specifically check this: https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/MoreExecutors.java#L559
i.e. guava does not cancel its wrapper future, when the wrapped delegate is canceled. Case like this makes me want to avoid extra wrapping when possible.

I've changed it to ApiFuture. But I still don't like it, because now it is harder to trace the calls and make some correctness guarantees.

garrettjonesgoogle replied:
oy, TODO's. 

In our case, if we always wrap with `ApiFuture`, can we mitigate the cancellation chaining problem?

vam-google replied:
*Note, all these problems existed with the previous RetriableCallable implementation (in a worse form, as activeFuture's cancelation was completelly ignored, plus the underlying call was made under the lock) , so this PR did not make things worse, just trying to mitigate the problems.

In general I don't know. There is still a problem for us, if somebody calls shutdownNow() on the underlying ScheduledExecutorService, which comes to the RetryingCallable from the outside. The problem here is that there are to many things to propagate in a reverse direction: the cancelation of java.util.concurrent executor's tasks, to guava's ListenableFutureTasks and then to RetryingFuture (plus all that custom lightweight wrappers we have). Also it requires support from the caller to actually even try to cancel the tasks returned by the shutdownNow() (most people don't do it). So multiple different layers of propagation in a reverse direction (from wrapped object to its wrapper), and as result information gets lost.

There are some heavyweight workarounds, which mitigate the issue (but still do not solve it completelly), like implementing custom shutdownNow on RetryExecutor, but that requires a lot of ugly stuff (now each task must know about RetryingFuture to cancel it when it gets canceled), and still does not solve the issue of somebody messing with aggregated pools/futures/tasks directly.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/grpc/RetryingCallable.java:
```diff
@@ -29,226 +29,151 @@
-          activeFuture = callFuture;
-        }
-      }
+      return null;
```
Something feels wrong about returning `null` from `call()`.

vam-google replied:
Yes it does feel wrong. It is essentially a hack, but when you have to make compatible two incompatible things, things like this happen. I still believe it is a very mild hack, so we can leave with it. 

The reason why this exist is same as described in the previous comment - the substitution of the future wrapped by RetryingFuture in the middle of the process - swapping ScheduledExecutorFutureTask with another asynchronous ApiFuture. 

In other words, what happens is chaining of **two asynchronous** calls, and since the second call is also asynchronous, the first one cannot return anything, because it does not know the actual operation result yet, and somebody has to wait for it to happen. The one who is waiting is the outside code, not the current intermediate task (otherwise it would consume threadpool thread).

garrettjonesgoogle commented on src/main/java/com/google/api/gax/retrying/ExceptionRetryAlgorithm.java:
```diff
@@ -0,0 +1,62 @@
+   * Returns {@code true} if another attempt should be made, or {@code false} otherwise.
+   *
+   * @param prevThrowable exception thrown by the previous attempt
+   * @return {@code true} if another attempt should be made, or {@code false} otherwise
```
You don't need the `@return` since it's redundant with the first sentence.

vam-google replied:
Removed. I wanted to keep the @return it as it feels more consistent and rigorous approach to documenting.

garrettjonesgoogle replied:
Makes sense, but if we want to solve that problem, we should probably apply it across the board, instead of just doing it occasionally (e.g. one person doing it for new methods, no one else doing it).

garrettjonesgoogle commented on src/main/java/com/google/api/gax/retrying/ExponentialRetryAlgorithm.java:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  /**
+   * Creates a first attempt {@link TimedAttemptSettings}. By default the first attempt is
```
Take off "by default", since there is no way to override it.

vam-google replied:
Removed

garrettjonesgoogle commented on src/main/java/com/google/api/gax/retrying/RetryAlgorithm.java:
```diff
@@ -0,0 +1,98 @@
+      Throwable prevThrowable, TimedAttemptSettings prevSettings) {
+    TimedAttemptSettings newSettings =
+        exceptionAlgorithm.createNextAttempt(prevThrowable, prevSettings);
+    if (prevSettings.equals(newSettings)) {
```
This approach means that the exception algorithm can't say that it wants the previous settings again and that it wants to take precedence over the timed algorithm. 

Maybe the exception algorithm could take the previous settings and and also the result from `timedAlgorithm.createNextAttempt(prevSettings)`, and it could decide what it wants to use?

vam-google replied:
I am against it. It essentially moves us back to what it was before this change - time and exception logics coupled together. If exception algorithm needs information from time alrorithm - this means it does care about timing more than it should. In such cases, I would encourage the clients to write a class which implements both TimedRetryAlgorithm and ExceptionRetryAlgorithm (the implementation may want to aggregate ExponentialRetryAlgorithm if it wants) - because what they need is the tightly coupled time-exception algorithm.

I think we reached good balance between flexibility and complexity of this. Doing more feels like overengineering.

garrettjonesgoogle replied:
It's fine to keep them separate; I was just trying to think of a way to allow the exception algorithm to unambiguously say that it has chosen what the timed attempt settings should be. Right now it is ambiguous.

vam-google replied:
In that case the exception algorithm should now about timed algorithm/any other algorithm. Now it does not.

garrettjonesgoogle commented on src/main/java/com/google/api/gax/retrying/RetryingExecutor.java:
```diff
@@ -0,0 +1,65 @@
+ *
+ * <ol>
+ * <li>Creating first attempt {@link RetryingFuture}, which acts as a facade, hiding from client
+ * code the actual scheduled retry attempts execution.
```
-> "the actual execution of scheduled retry attempts"

vam-google replied:
Changed

garrettjonesgoogle commented on src/main/java/com/google/api/gax/retrying/RetryingExecutor.java:
```diff
@@ -0,0 +1,65 @@
+   *
+   * @param retryingFuture the future previously returned by {@link #createFuture(Callable)}
+   */
+  void submit(RetryingFuture<ResponseT> retryingFuture);
```
It's not clear from the docs if a new retrying future needs to be created for every attempt, or if the same one should be re-used, or if it doesn't matter.

shinfan commented on src/main/java/com/google/api/gax/core/RetrySettings.java:
```diff
@@ -189,6 +206,13 @@ public Builder toBuilder() {
     public abstract double getRetryDelayMultiplier();
 
     /**
+     * MaxAttempts defines the maximum number of attempts to perform. If number of attempts reaches
```
If *the* number

shinfan commented on src/main/java/com/google/api/gax/grpc/RetryingCallable.java:
```diff
@@ -29,226 +29,151 @@
+      CallContext oldCtx, Duration rpcTimeout) {
+    CallOptions oldOpt = oldCtx.getCallOptions();
+    CallOptions newOpt = oldOpt.withDeadlineAfter(rpcTimeout.getMillis(), TimeUnit.MILLISECONDS);
+    CallContext newCtx = oldCtx.withCallOptions(newOpt);
```
Sorry I did not notice this in my previous review. To be consistent, we should avoid using such abbreviations (e.g. context vs ctx)

vam-google replied:
Yes, though it was an old piece of code, I did not write it, just moved. I renamed the variables and arguments.

shinfan commented on src/main/java/com/google/api/gax/retrying/TimedAttemptSettings.java:
```diff
@@ -0,0 +1,107 @@
+/**
+ * Timed attempt execution settings. Defines time-specific properties of a retry attempt.
+ */
+public class TimedAttemptSettings {
```
`TimeAttemptSettings` gives me the impression that it is associated with `TimedRetryAlgorithm` but it is also used by `ExceptionRetryAlgorithm` as well. Maybe rename it to just `AttemptSettings` or `RetryAttemptSettings`?

vam-google replied:
RetryAttemptSettings was the old name of it, which was not liked...

The important part is that TimeAttemptSettings is actually associated with TimedRetryAlgorithm much more than with ExceptionRetryAlrogithm, so your impression is not far from being correct. For example the gcj version of it ignores the timedretrysettings parameter. Gax uses it, because of some older design decision.

shinfan replied:
Okay, fair enough.

shinfan commented on src/main/java/com/google/api/gax/retrying/ScheduledRetryingExecutor.java:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  /**
+   * Submits an attempt for execution in a different thread.
```
Add a comment here to indicate if use a direct scheduler, the execution will be in the same thread.

vam-google replied:
I decided to remove it (I had it before) because, for example MoreExecutors has the "directExecutor()" version for a regular executor, but does not have one for ScheduledExecutorService. Probably because "scheduling" within same thread does not make much sense (it is rather "delaying" than "scheduling"). So I removed the notion of direct executor, because DirectScheduledExecutor is not a thing.

shinfan commented on src/main/java/com/google/api/gax/grpc/RetryingCallable.java:
```diff
@@ -29,226 +29,151 @@
+        new GrpcRetryCallable<>(callable, request, context);
+
+    RetryingFuture<ResponseT> retryingFuture = scheduler.createFuture(retryCallable);
+    retryCallable.setExternalFuture(retryingFuture);
```
This looks circular... so the scheduler takes a RetryCallable to create a RetryingFuture and you set the RetryingFuture to the RetryCallable.

vam-google replied:
You are right, it is circular and it is designed to be circular. This is forced by the current gax FutureCallables design. Please check my previous comment to this PR for more details what is happening here and why: https://github.com/googleapis/gax-java/pull/235#discussion_r106738328

(also the previous implementation of RetryingCallable was circular as well).

vam-google commented:
Adressed PR comments, PTAL.

vam-google commented:
PTAL

garrettjonesgoogle commented:
Offline agreements: 

1. Rename `SystemClock` to `CurrentMillisClock`
2. Have `ExceptionAlgorithm.createNextAttempt` return null if it is not overriding the settings

LGTM after these are done - Shin should probably also approve.

vam-google commented:
PTAL

shinfan commented on src/main/java/com/google/api/gax/retrying/ScheduledRetryingExecutor.java:
```diff
@@ -0,0 +1,102 @@
+
+/**
+ * The retry executor which uses {@link ScheduledExecutorService} to schedule an attempt tasks.
+ * Unless a direct executor service is used, this handler will schedule attempts for an execution in
```
Remove this as well.

vam-google replied:
Removed

vam-google commented:
Done

shinfan commented:
LGTM
